### PR TITLE
Locks stored as annotations are safe to write

### PR DESCRIPTION
**Pull request by gforcada**
plone.protect safeWrite looks for the oid of the object passed.

self.context has one oid and the locks variable which ends up as an annotation on self.context has a different oid, thus causing a not protected write on read.

@vangheem I spend like two or three days staring at the code until the [excellent blog post](http://devblog.4teamwork.ch/blog/2015/10/12/debugging-csrf-protection-false-positives-in-plone/) from @lukasgraf saved my day.

Does this look fine? I'm seeing that at every request this ``lock`` method is called twice and the first time ``locks`` does not have a ``_p_oid`` so a warning by ``safeWrite`` will be generated...

Last thing, sorry for dumping my notes here, I see that on the [previous commit](https://github.com/plone/plone.locking/commit/e23cd15326a87cd224b7d62a0efd781246ab6cb2) that fixed the CSRF issues in this ``lock`` method self.context is both ``self.context._v_safe_write = True`` and ``safeWrite(self.context)`` I didn't dig further but given the semantics seems that they are doing the same somehow (making self.context safe for writing).

### CODE REVIEW

**Comment by lukasgraf** [comment]
The volatile `_v_safe_write` attribute was the first attempt at whitelisting objects, and has been [superseded by the `safeWrite()` function](https://github.com/plone/plone.protect/commit/98e2a3a2d86b6c22051e058f6cd1cc7752e8da66), so `safeWrite()` should be enough.

As for your change, I'm a bit confused about why [this existing whitelisting of the context's annotation doesn't seem to work for you](https://github.com/plone/plone.locking/blob/532a94ec445addbefa28b877ac685c962805250f/plone/locking/lockable.py#L166). The only way I could see that the annotations returned by `TTWLockable._locks()` annotations haven't been whitelisted yet is with `create=False`, where it would hit the final `return {}` - but that's a plain dictionary, not a `PersistentDict()`, so that shouldn't trigger a DB write.

**Comment by gforcada** [comment]
@lukasgraf thanks, again, for the pointer. I put a pdb at that plone.locking safeWrite call as well as in plone.protect code and this is what I see:

**On plone.locking**
```
> /home/gil/Documents/freitag/git/zope/src/plone.locking/plone/locking/lockable.py(162)_locks()
    161             import ipdb; ipdb.set_trace()
--> 162             safeWrite(annotations.obj.__annotations__)
    163         except AttributeError:

ipdb> from ZODB.utils import u64
ipdb> hex(u64(annotations.obj.__annotations__._p_oid)) ; [f for f in annotations.obj.__annotations__.items()]
'0xbcde5964ead15b'
[OTHER_DATA... ('plone.locking', {u'plone.locking.stealable': {'token': '0.897978962067-0.630613757163-00105A989226:1445284265.561', 'type': <plone.locking.interfaces.LockType object at 0x7f70ae228710>}})]

**On plone.protect**
```
> /home/gil/Documents/freitag/git/zope/src/plone.protect/plone/protect/auto.py(218)_check()
    217                         import ipdb; ipdb.set_trace()
--> 218                         break
    219                 if not safe:

ipdb> obj
{u'plone.locking.stealable': {'token': '0.897978962067-0.630613757163-00105A989226:1445284265.561', 'type': <plone.locking.interfaces.LockType object at 0x7f70ae228710>}}
ipdb> from ZODB.utils import u64; hex(u64(obj._p_oid))
'0xbcde5964ead296'
ipdb> safe_oids
['\x00\xbc\xdeYd\xea\xd1[', '\x00\xbc\xdeYd\xea\xd2\x94', '\x00\xbc\xdeYd\xea\xd0\xf9']
ipdb> '0xbcde5964ead15b' in [hex(u64(f)) for f in safe_oids]  # ANNOTATION _p_oid
True
ipdb> '0xbcde5964ead296' in [hex(u64(f)) for f in safe_oids]  # _p_oid OF THE OBJECT THAT RAISES THE CSRF
False
```

So basically that the ``safeWrite`` on the annotation is for the annotation object not for its related objects (being a BTree).

On plone.locking we can see which ``_p_oid`` has the annotation and on the last check on plone.protect we can see that that ``_p_oid`` is on the list of ``safe_oids``.

The object that gives the problem though (on plone.protect) has a different ``_p_oid`` though.

Maybe in some cases is working and others (like mine) not, because of how the BTrees allocate their data? So in my case that annotation already has some data on it prior to add the locking info (specifically these keys: ``'plone.contentrules.localassignments', 'plone.folder.ordered.order', 'plone.folder.ordered.pos'``).

Does that make sense at all?

**Comment by lukasgraf** [comment]
Ah, now I understand: The modified object isn't the annotations themselves, but a value inside them (i.e. the mapping at `annotations[ANNOTATION_KEY]`).

So now your change makes sense to me :+1:

However, I would try if you can maybe do a check like `if hasattr(locks, '_p_oid'): safeWrite(locks)` so that no annoying bogus warnings are produced.

**Comment by gforcada** [comment]
@lukasgraf thanks for the review.

I amended the commit and did a ``getattr`` (a hasattr can swallow exceptions).

Let's see what jenkins has to say about it: http://jenkins.plone.org/job/pull-request-5.0/485

**Comment by vangheem** [comment]
It's fine. Merge whenever.

**Comment by jensens** [comment]
here the changelog entry is missing!

**Comment by gforcada** [comment]
I added it after merging, sorry: https://github.com/plone/plone.locking/commit/d40d563a8dcdfeca0c681d3b8b8d8b952951d88f
